### PR TITLE
Move IP checks to startup time

### PR DESCRIPTION
**Pull request by Viktor-Yakovchuk**
Move validation of static IP addresses from setting the config to startup.
It allows to copy a container that has one or more static IP addresses:

fix issue: #10114 

Also, it allows setting up local already used IP to container by `lxc config device set`

### CODE REVIEW

**Comment by Viktor-Yakovchuk** [comment]
jenkins: ok to test

**Comment by Viktor-Yakovchuk** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
jenkins: test this please

**Review by tomponline — CHANGES_REQUESTED** [review]
Thanks! I'm running the tests for this now.
But we will want to add some tests for this change of logic to ensure the issue it fixes is handled.

**Comment by Viktor-Yakovchuk** [comment]
I watched through the tests and didn't find what exactly fails with my patch.
As I understand, logs problems are here :
+ LXD_DIR=/lxc-ci/build/tmp.AGgBAsvFob/go/src/github.com/lxc/lxd/test/tmp.Zir/mh7 lxc config device set c2 eth0 ipv4.address=192.0.2.2
+ LXC_LOCAL=1 lxc_remote config device set c2 eth0 ipv4.address=192.0.2.2
+ set +x
+ eval /lxc-ci/build/tmp.AGgBAsvFob/go/bin/lxc "config" "device" "set" "c2" "eth0" "ipv4.address=192.0.2.2" --verbose
+ /lxc-ci/build/tmp.AGgBAsvFob/go/bin/lxc config device set c2 eth0 ipv4.address=192.0.2.2 --verbose
+ false
+ cleanup

File I was looking is https://jenkins.linuxcontainers.org/job/lxd-github-pull-test/6932/arch=amd64,backend=dir,compiler=golang-1.18,section=cluster/consoleText
But I can't find where the code that responsible for that test. I have tried to look it in lxd/test and in lxc-ci. It is complicated question for me now dew to lack knowledge haw exactly this CI works.

**Review by tomponline — CHANGES_REQUESTED** [review]
Thanks for this!

As per @stgraber comment (https://github.com/lxc/lxd/issues/10114#issuecomment-1077584756): 

> I'd prefer we move those checks to startup time when possible. It may mean we need to move some of the dnsmasq logic to occur on first start instead then.

We also need to ensure that the `bridged` NIC's `Add()` function is no longer implemented (as it would rebuild dnsmasq static config and with your change that static config is no longer validated at add time, and so we cannot reliably add dnsmasq static config until first start).

The question is how to define and detect an instance device's first start?

I think at least for now (as only `bridged` NICs actually use the `Add()` function) we implement this detection inside the `bridged` NIC's `PreStart()` function and check if the device specific dnsmasq config file exists, if it doesn't then we can consider this its first start and regenerate the dnsmasq config.

File: test/suites/basic.sh [hunk 1]
```diff
@@ -126,7 +126,11 @@ test_basic_usage() {
 
   # Test container copy
   lxc copy bar foo
+  lxc config device set foo eth0 ipv4.address=192.0.2.2
+  # Test container with static IP copy
+  lxc copy foo baz
```
Comment (tomponline) [inline test/suites/basic.sh:131]: Can you try starting it here and check it fails too.

File: test/suites/basic.sh [hunk 2]
```diff
@@ -126,7 +126,11 @@ test_basic_usage() {
...
   lxc delete foo
+  lxc delete baz
```
Comment (tomponline) [inline test/suites/basic.sh:134]: Can you add a test for restoring a snapshot that has a static IP assigned to a NIC that conflicts with another instance.
This checks for the first scenario described in https://github.com/lxc/lxd/issues/10114#issuecomment-1077449896

**Comment by tomponline** [comment]
We can then also get rid of the `Add()` from the `device` package interface and the `deviceAdd` functions in instance drivers.

**Comment by Viktor-Yakovchuk** [comment]
I'm sorry, but I'll be able to continue working on the pull request in a few days

**Comment by stgraber** [comment]
Converted the PR to draft.

File: test/suites/basic.sh (hunk 1 above)
Comment (Viktor-Yakovchuk) [reply test/suites/basic.sh:131]: I check this in [clustering.sh](https://github.com/lxc/lxd/pull/10363/commits/7b696ff1dee5e5e409b4cf0a1845c7727ecbf6cb#diff-7cd3b6dcbb2b188abfdc6a6141d12b1f779cd4ce27356d86f1e855dd1414923a)

File: test/suites/basic.sh (hunk 2 above)
Comment (Viktor-Yakovchuk) [reply test/suites/basic.sh:134]: Done

**Comment by Viktor-Yakovchuk** [comment]
I searched for the [deviceAdd](https://github.com/lxc/lxd/blob/ae0601a1b59783a7c818247200ca64de26df8278/lxd/instance/drivers/driver_lxc.go#L1380) function
And it used twice [here](https://github.com/lxc/lxd/blob/ae0601a1b59783a7c818247200ca64de26df8278/lxd/instance/drivers/driver_lxc.go#L306) and [here](https://github.com/lxc/lxd/blob/ae0601a1b59783a7c818247200ca64de26df8278/lxd/instance/drivers/driver_lxc.go#L4785)
As I understand Add() function uses in 20 files.

As to Add() function, I delete it from [nic_bridged.go](https://github.com/lxc/lxd/pull/10363/commits/301cc026c95cc4dba65a43c22e26521c1561d93a#diff-4852ea74f7ad2af08879be6d9495a75b0e58f814d9f9eb1f82e74485a9a3b7ba)

**Comment by tomponline** [comment]
> I searched for the [deviceAdd](https://github.com/lxc/lxd/blob/ae0601a1b59783a7c818247200ca64de26df8278/lxd/instance/drivers/driver_lxc.go#L1380) function And it used twice [here](https://github.com/lxc/lxd/blob/ae0601a1b59783a7c818247200ca64de26df8278/lxd/instance/drivers/driver_lxc.go#L306) and [here](https://github.com/lxc/lxd/blob/ae0601a1b59783a7c818247200ca64de26df8278/lxd/instance/drivers/driver_lxc.go#L4785) As I understand Add() function uses in 20 files.
> 
> As to Add() function, I delete it from [nic_bridged.go](https://github.com/lxc/lxd/pull/10363/commits/301cc026c95cc4dba65a43c22e26521c1561d93a#diff-4852ea74f7ad2af08879be6d9495a75b0e58f814d9f9eb1f82e74485a9a3b7ba)

I think we can remove Add from the interface and drivers entirely.

**Comment by Viktor-Yakovchuk** [comment]
I removed [deviceAdd](https://github.com/lxc/lxd/blob/ae0601a1b59783a7c818247200ca64de26df8278/lxd/instance/drivers/driver_lxc.go#L1380) and its calls. I also notice that after it there are such calls:
`revert.Add(func() { _ = d.deviceRemove(dev, false) })`
Should we also remove deviceRemove function?

**Review by tomponline — CHANGES_REQUESTED** [review]
Getting there, though still some missing parts.

**Comment by tomponline** [comment]
@Viktor-Yakovchuk are you still able to progress with this? Thanks

**Comment by Viktor-Yakovchuk** [comment]
Sorry for delay, I will push new version today

File: test/suites/snapshots.sh [hunk 3]
```diff
@@ -197,6 +197,27 @@ snap_restore() {
 
   ##########################################################
 
+  # test restore with static IP
+  lxc launch testimage baz
+  lxc launch testimage qux
+
+  lxc config device override baz eth0 ipv4.address=10.93.225.101
```
[the remark this answers is not in the document]
Comment (Viktor-Yakovchuk) [reply test/suites/snapshots.sh]: Fixed

File: lxd/device/nic_bridged.go [hunk 4]
```diff
@@ -367,17 +367,6 @@ func (d *nicBridged) UpdatableFields(oldDevice Type) []string {
 	return []string{"limits.ingress", "limits.egress", "limits.max", "ipv4.routes", "ipv6.routes", "ipv4.routes.external", "ipv6.routes.external", "ipv4.address", "ipv6.address", "security.mac_filtering", "security.ipv4_filtering", "security.ipv6_filtering"}
 }
 
-// Add is run when a device is added to a non-snapshot instance whether or not the instance is running.
-func (d *nicBridged) Add() error {
```
[the remark this answers is not in the document]
Comment (Viktor-Yakovchuk) [reply lxd/device/nic_bridged.go:471]: Done

File: lxd/instance/drivers/driver_qemu.go [hunk 5]
```diff
@@ -4766,21 +4747,6 @@ func (d *qemu) updateDevices(removeDevices deviceConfig.Devices, addDevices devi
 			d.logger.Error("Device stop validation failed", logger.Ctx{"devName": dd.Name, "err": err})
 		}
 
-		// If a device was returned from deviceLoad even if validation fails, then try and stop and remove.
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:4769]: We still need this section.

File: lxd/instance/drivers/driver_qemu.go [hunk 6]
```diff
@@ -5073,14 +5026,6 @@ func (d *qemu) Delete(force bool) error {
 				// valid devices to be remove.
 				d.logger.Error("Device remove validation failed", logger.Ctx{"device": entry.Name, "err": err})
 			}
-
-			// If a device was returned from deviceLoad even if validation fails, then try and remove.
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:5077]: We still need this section.

File: lxd/instance/drivers/driver_qemu.go [hunk 7]
```diff
@@ -5121,28 +5066,6 @@ func (d *qemu) Delete(force bool) error {
...
-	l.Debug("Adding device")
-
-	if instanceRunning && !dev.CanHotPlug() {
-		return fmt.Errorf("Device cannot be added when instance is running")
-	}
-
-	return dev.Add()
-}
-
-func (d *qemu) deviceRemove(dev device.Device, instanceRunning bool) error {
```
Comment (tomponline) [inline lxd/instance/drivers/driver_qemu.go:5135]: We still need this function.

**Review by tomponline — CHANGES_REQUESTED** [review]
I think there has been some misunderstanding.
We still need deviceRemove, its just that we don't need to call deviceRemove as a revert when not calling deviceAdd anymore.

File: lxd/device/device_interface.go [hunk 8]
```diff
@@ -62,9 +58,6 @@ type Device interface {
 	// either due to unplugging it from a running instance or instance is being shutdown.
 	// Returns run-time configuration needed for detaching the device from the instance.
 	Stop() (*deviceConfig.RunConfig, error)
-
-	// Remove performs any host-side cleanup when a device is removed from an instance.
-	Remove() error
```
[the remark this answers is not in the document]
Comment (Viktor-Yakovchuk) [reply lxd/device/device_interface.go]: I brought it back

File: lxd/instance/drivers/driver_qemu.go (hunk 5 above)
Comment (Viktor-Yakovchuk) [reply lxd/instance/drivers/driver_qemu.go]: I brought it back here and in `lxd/instance/drivers/driver_lxc.go`

File: lxd/instance/drivers/driver_qemu.go (hunk 6 above)
Comment (Viktor-Yakovchuk) [reply lxd/instance/drivers/driver_qemu.go]: I brought it back here and in `lxd/instance/drivers/driver_lxc.go`

File: lxd/instance/drivers/driver_qemu.go (hunk 7 above)
Comment (Viktor-Yakovchuk) [reply lxd/instance/drivers/driver_qemu.go:5135]: I brought it back

File: lxd/device/nic_bridged.go [hunk 9]
```diff
@@ -382,6 +382,19 @@ func (d *nicBridged) PreStartCheck() error {
 	// Check there isn't another NIC with any of the same addresses specified on the same cluster member.
 	// Can only validate this when the instance is supplied (and not doing profile validation).
 	if d.inst != nil {
+
+		deviceStaticFileName := dnsmasq.StaticAllocationFileName(d.inst.Project(), d.inst.Name(), d.Name())
```
Comment (tomponline) [inline lxd/device/nic_bridged.go:386]: Please can you move this after the check below described in the comment above about "Check there isn't another NIC with any of the same addresses specified on the same cluster member." as we shouldn't build the NICs config if it potentially has conflicts with existing config. So move this section to the end of the function please.

Comment (Viktor-Yakovchuk) [reply lxd/device/nic_bridged.go:386]: Done

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
@stgraber jenkins wont run on this PR:

```
Receiving objects:  76% (131128/172536), 59.47 MiB | 29.73 MiB/s
error: RPC failed; curl 56 GnuTLS recv error (-54): Error in the pull function.
fatal: the remote end hung up unexpectedly
fatal: early EOF
fatal: index-pack failed
```

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
All tests failing.

**Comment by Viktor-Yakovchuk** [comment]
Thanks, I fixed. Problem was in my new test.

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by Viktor-Yakovchuk** [comment]
I moved test with static ip from basic.sh to container_devices_nic_bridged.sh because there is ready network configuration for containers in it.

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by Viktor-Yakovchuk** [comment]
I will try to reproduce it on my local machine tomorrow.

**Comment by tomponline** [comment]
enkins: test this please

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
Tests still failing.

If you're happy with it, I can take a look at fixing this up next week?

**Comment by Viktor-Yakovchuk** [comment]
I am trying to launch tests on my local machine and receive following error:
```
+ lxc_remote remote add test images.linuxcontainers.org
+ set +x
+ eval /root/go/bin/lxc "remote" "add" "test" "images.linuxcontainers.org" --verbose
+ /root/go/bin/lxc remote add test images.linuxcontainers.org --verbose
Certificate fingerprint: 4ba8bc8fe867277ce1ddc4f1eb80ad3cb0548243f7e930850744ceead0e306d7
ok (y/n/[fingerprint])? y
Error: Failed to fetch https://uk.lxd.images.canonical.com/1.0: 404 Not Found
```

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
@Viktor-Yakovchuk please don't spend any more time on this, as after some discussions with @stgraber  around how OVN NICs operate, it has become clear that the way we were planning to relax the NIC add-time validation will introduce further issues by preventing a NIC from reserving its static IP at add time. I'm working on a new PR for this, and I'll try and cherry-pick as much from this PR as possible. Thanks

**Comment by Viktor-Yakovchuk** [comment]
I would like to make some small improvement of last commit. 

I am also trying to launch test on my local machine. I am on my branch localIPfix (from this branch I create pull request)
I commented `run_test test_static_analysis "static analysis"` in test/main.sh
After I go to test dir  and launch tests by calling ./main.sh I revived following error:
```
+ lxc_remote remote add test images.linuxcontainers.org
+ set +x
+ eval /root/go/bin/lxc "remote" "add" "test" "images.linuxcontainers.org" --verbose
+ /root/go/bin/lxc remote add test images.linuxcontainers.org --verbose
Certificate fingerprint: 4ba8bc8fe867277ce1ddc4f1eb80ad3cb0548243f7e930850744ceead0e306d7
ok (y/n/[fingerprint])? y
Error: Failed to fetch https://uk.lxd.images.canonical.com/1.0: 404 Not Found
```

**Comment by tomponline** [comment]
Closed in place of https://github.com/lxc/lxd/pull/10546

**Pull request closed without merging**